scripts/tls.py

```diff
@@ -340,13 +349,78 @@ def clientCmd(argv):
         sys.exit(-1)
     printGoodConnection(connection, stop-start)
     printExporter(connection, expLabel, expLength)
+    session = connection.session
+    connection.send("GET / HTTP/1.0\r\n\r\n")
+    while True:
+        try:
+            r = connection.recv(10240)
+            if not r:
+                break
+        except socket.timeout:
+            break
+        except TLSAbruptCloseError:
+            break
+    connection.close()
+    # we're expecting an abrupt close error which marks the session as
+    # unreasumable, override it
+    session.resumable = True
+
+    print("Received {0} ticket[s]".format(len(connection.tickets)))
+    assert connection.tickets is session.tickets
```

Comment:
Note that this assertion triggers when connecting to nearly any HTTPS server on the internet.

$ tls.py client www.yahoo.com:443
Handshake success
  Handshake time: 0.181 seconds
  Version: TLS 1.2
  Cipher: aes128gcm python
  Ciphersuite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
  No client certificate provided by peer
  Server X.509 SHA1 fingerprint: 3914af80223c833f26df001cbf342eff8a31aba1
  Key exchange signature: rsa_pkcs1_sha512
  Group used for key exchange: secp256r1
  SNI: www.yahoo.com
  Next-Protocol Negotiated: None
  Encrypt-then-MAC: False
  Extended Master Secret: False
Received 0 ticket[s]
Traceback (most recent call last):
  File "/usr/local/bin/tls.py", line 541, in <module>
    clientCmd(sys.argv[2:])
  File "/usr/local/bin/tls.py", line 377, in clientCmd
    assert connection.tickets is session.tickets
AssertionError

$ tls.py client www.amazon.com:443
Handshake success
  Handshake time: 0.186 seconds
  Version: TLS 1.2
  Cipher: aes128gcm python
  Ciphersuite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
  No client certificate provided by peer
  Server X.509 SHA1 fingerprint: 0e6bdcae568e3613c6d173c9b112c357ff5967b2
  Key exchange signature: rsa_pkcs1_sha512
  Group used for key exchange: secp256r1
  SNI: www.amazon.com
  Next-Protocol Negotiated: None
  Encrypt-then-MAC: False
  Extended Master Secret: False
Received 0 ticket[s]
Traceback (most recent call last):
  File "/usr/local/bin/tls.py", line 541, in <module>
    clientCmd(sys.argv[2:])
  File "/usr/local/bin/tls.py", line 377, in clientCmd
    assert connection.tickets is session.tickets
AssertionError

$ tls.py client www.google.com:443
Handshake success
  Handshake time: 0.068 seconds
  Version: TLS 1.2
  Cipher: chacha20-poly1305 python
  Ciphersuite: TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
  No client certificate provided by peer
  Server X.509 SHA1 fingerprint: 62231769ed0bf01ecce073c24a668578c5ffb66e
  Key exchange signature: rsa_pss_rsae_sha256
  Group used for key exchange: x25519
  SNI: www.google.com
  Next-Protocol Negotiated: None
  Encrypt-then-MAC: False
  Extended Master Secret: True
Received 0 ticket[s]
Traceback (most recent call last):
  File "/usr/local/bin/tls.py", line 541, in <module>
    clientCmd(sys.argv[2:])
  File "/usr/local/bin/tls.py", line 377, in clientCmd
    assert connection.tickets is session.tickets
AssertionError

Reply:
please file an issue

Reply:
ACK https://github.com/tomato42/tlslite-ng/issues/317